File: Python/treehopper/libraries/displays/LedDriver.py
import logging
import math
from treehopper.libraries.Flushable import Flushable
from treehopper.libraries.displays.Led import Led
import abc
from treehopper.utils.utils import constrain


class LedDriver(Flushable):
    def __init__(self, num_leds: int, has_global_brightness_control: bool, has_individual_brightness_control: bool):
        Flushable.__init__(self)
        self._logger = logging.getLogger(__name__)
        self._brightness = 1.0
        self.global_brightness = has_global_brightness_control
        self.individual_brightness = has_individual_brightness_control
        self.leds = []
        for i in range(num_leds):
            self.leds.append(Led(self, i, self.individual_brightness))

    @abc.abstractmethod
    def flush(self, force=False):
        pass

    @abc.abstractmethod
    def _led_state_changed(self, led: Led):
        pass

    @abc.abstractmethod
    def _led_brightness_changed(self, led: Led):
        pass

    @abc.abstractmethod
    def _set_global_brightness(self, value: float):
        pass

    @property
    def brightness(self, value):
        if math.isclose(value, self.brightness):
            return

        if value < 0.0 or value > 1.0:
            self._logger.warning("Constraining out-of-range brightness value to 0.0-1.0")

        self._brightness = constrain(value, 0.0, 1.0)

        if self.global_brightness:
            self._set_global_brightness(self._brightness)
        else:
            saved_autoflush = self.auto_flush
            self.auto_flush = False
            for led in self.leds:
                led.brightness = self._brightness

            self.flush()
            self.auto_flush = saved_autoflush


    @brightness.getter
    def brightness(self):
        return self._brightness

File: Java/api/treehopper.libraries/src/main/java/io/treehopper/libraries/displays/LedDriver.java
package io.treehopper.libraries.displays;

import io.treehopper.libraries.IFlushable;

import java.util.ArrayList;
import java.util.List;

/**
 * Base ILedDriver implementation
 */
public abstract class LedDriver implements ILedDriver, IFlushable {


    private boolean autoFlush;
    private double brightness;
    private List<Led> leds;


    public LedDriver(int numLeds, boolean HasGlobalBrightnessControl, boolean HasIndividualBrightnessControl) {
        this.autoFlush = true;
        this.brightness = 0.0;
        this.leds = new ArrayList<>();
        for (int i = 0; i < numLeds; i++) {
            Led led = new Led(this, i, HasIndividualBrightnessControl);
            this.leds.add(led);
        }
    }

    public List<Led> getLeds() {
        return leds;
    }

    @Override
    public boolean isAutoFlushEnabled() {
        return autoFlush;
    }

    public void setAutoFlushEnabled(boolean autoFlush) {
        this.autoFlush = autoFlush;
    }


    @Override
    public abstract boolean hasGlobalBrightnessControl();

    @Override
    public abstract boolean hasIndividualBrightnessControl();

    public double getBrightness() {
        return brightness;
    }

    public void setBrightness(double brightness) {
        if (Math.abs(this.brightness - brightness) < 0.0005) return;
        if (this.brightness < 0 || brightness > 1)
            throw new RuntimeException("Valid brightness is from 0 to 1");

        this.brightness = brightness;

        _setBrightness(brightness);
    }

    protected abstract void _setBrightness(double brightness);

    public abstract void ledStateChanged(Led led);

    public abstract void ledBrightnessChanged(Led led);

    public void clear() {
        boolean autoflushSave = isAutoFlushEnabled();
        setAutoFlushEnabled(false);
        for (Led led : this.leds) {
            led.setState(false);
        }
        flush(true);
        setAutoFlushEnabled(autoflushSave);
    }

    public abstract void flush(boolean force);

}
